more/pixels/gallery.js

```diff
@@ -79,227 +56,154 @@ export function makeGallery(
-        transferRightAssay,
-      );
-
-      const transferRightPurse = transferRightMint.mint(transferAmount);
-      const useRightPurse = useRightMint.mint(useAmount);
-
-      const transferRightPayment = await transferRightPurse.withdrawAll(
-        'transferRights',
-      );
-      const useRightPayment = await useRightPurse.withdrawAll('useRights');
+  function setPixelListState(pixelList, newColor) {
+    for (let i = 0; i < pixelList.length; i += 1) {
+      const pixel = pixelList[i];
+      state[pixel.x][pixel.y] = newColor;
+      // eslint-disable-next-line no-use-before-define
+      lruQueue.requeue(pixel);
+    }
 
-      return {
-        transferRightPayment,
-        useRightPayment,
-      };
-    });
+    // for now we pass the whole state, but only once per call
+    stateChangeHandler(getState());
   }
 
-  // merge UseRights and TransferRights into a pixel
-  async function toPixel(transferRightPaymentP) {
-    return Promise.resolve(transferRightPaymentP).then(
-      async transferRightPayment => {
-        // someone else may have the useRightPayment so we must destroy the
-        // useRight
-
-        // we have an exclusive on the transfer right
-        const transferAmount = transferRightPayment.getBalance();
-        await transferRightIssuer.getExclusiveAll(transferRightPayment);
-
-        const pixelListAmount = transferRightAssay.toPixel(
-          transferAmount,
-          pixelAssay,
-        );
-
-        const { useAmount } = pixelAssay.toTransferAndUseRights(
-          pixelListAmount,
-          useRightAssay,
-          transferRightAssay,
-        );
-
-        // commit point
-        await useRightMint.destroy(useAmount);
-        await transferRightMint.destroy(transferAmount);
-
-        const pixelPayment = await gallerySplitPixelPurse.withdraw(
-          pixelListAmount,
-          'pixels',
-        ); // conserve pixels
-        return pixelPayment;
-      },
-    );
+  // anyone can getPixelColor, no restrictions, no tokens
+  function getPixelColor(x, y) {
+    const rawPixel = { x: Nat(x), y: Nat(y) };
+    return state[rawPixel.x][rawPixel.y];
   }
 
-  function insistColor(_myColor) {
-    // TODO: check whether allowed
+  function makeUseObj(issuer, asset) {
+    const useObj = harden({
+      changeColor(amount, newColor) {
+        insistNonEmptyAmount(issuer, amount);
+        insistAssetHasAmount(issuer, asset, amount);
+        insistColor(newColor);
+        const pixelList = getPixelList(issuer, amount);
+        setPixelListState(pixelList, newColor);
+        return amount;
+      },
+      changeColorAll(newColor) {
+        return useObj.changeColor(asset.getBalance(), newColor);
+      },
+      getRawPixels() {
+        const assay = issuer.getAssay();
+        const pixelList = assay.quantity(asset.getBalance());
+        return pixelList;
+      },
+      getColors() {
+        const pixelList = useObj.getRawPixels();
+        const colors = [];
```

Comment:
The design challenge here is that we can't return a set because it (as far as I know) can't be serialized easily. I just pushed up a change that still returns an array, but the array contains objects with x, y, and color properties so at least the color is mapped to a pixel. Do you know if there's a better way?

Reply:
I think an interface that explicitly reveals which color applies to each pixel is a fine answer.